route_guide/route_guide_server.cc

```diff
@@ -55,166 +66,259 @@ float GetDistance(const Point& start, const Point& end) {
+                   } else {
+                     // The invalid request indicates the client has finished
+                     // sending all its requests. We can send the final
+                     // response.
+                     system_clock::time_point end_time = system_clock::now();
+                     RouteSummary summary;
+                     summary.set_point_count(point_count);
+                     summary.set_feature_count(feature_count);
+                     summary.set_distance(static_cast<long>(distance));
+                     auto secs = std::chrono::duration_cast<
+                         std::chrono::seconds>(
+                         end_time - start_time);
+                     summary.set_elapsed_time(secs.count());
+                     call.WriteAndFinish(summary, Status::OK);
+                   }
+                 });
+           }))
+        | Loop();
+  }
 
-    status = server->Serve<RouteGuide, Stream<RouteNote>, Stream<RouteNote>>(
-        "RouteChat",
-        [this](auto* call, auto&& note) {
-          if (note) {
-            std::unique_lock<std::mutex> lock(mu_);
-            for (const RouteNote& n : received_notes_) {
-              if (n.location().latitude() == note->location().latitude() &&
-                  n.location().longitude() == note->location().longitude()) {
-                call->Write(n);
-              }
-            }
-            received_notes_.push_back(*note);
-          } else {
-            call->Finish(Status::OK);
-          }
-        },
-        [](auto*, bool) {});
-
-    return status;
+  // ServeRouteChat demonstrates a bidirectional streaming RPC. The client
+  // sends a sequence of RouteNote messages (see protos/route_guide.proto),
+  // which each have a location. The server replies with any RouteNotes that it
+  // has previously received for those locations.
+  auto ServeRouteChat(Server* server) {
+    return server->Accept<
+               RouteGuide,
+               Stream<RouteNote>,
+               Stream<RouteNote>>("RouteChat")
+        | Map(Then([this](auto&& context) {
+             return Server::Handler(std::move(context))
+                 .body([this](auto& call, auto&& note) {
+                   // Check if the request is valid.
+                   if (note) {
+                     for (const RouteNote& n : received_notes_) {
+                       if (n.location().latitude()
+                               == note->location().latitude()
+                           && n.location().longitude()
+                               == note->location().longitude()) {
+                         // Response to the client.
+                         call.Write(n);
+                       }
+                     }
+                     received_notes_.push_back(*note);
+                   } else {
+                     // When writes are done just finish call.
+                     call.Finish(Status::OK);
+                   }
```

Comment:
We can increase readability by reducing the indentation of this code, by using `return`:
```suggestion
                   if (note == nullptr) {
                     // When writes are done just finish call.
                     call.Finish(Status::OK);
                     return;
                   }
                   // We received a valid note. Reply with other notes
                   // from the same location.
                   for (const RouteNote& n : received_notes_) {
                     if (n.location().latitude()
                             == note->location().latitude()
                         && n.location().longitude()
                             == note->location().longitude()) {
                       // Response to the client.
                       call.Write(n);
                     }
                   }
                   received_notes_.push_back(*note);
```
The same goes in a few other places in this PR where we're checking for request/response validity.

Reply:
Fully agreed, great idea RJ:)